### PR TITLE
Move Towards Generic API

### PR DESCRIPTION
**Pull request by sjudson**
At present, we have a variety of pipelines for each `frontend` -> `vm` -> `proving scheme` implementation.

- `tools` (minus `src/command/jolt.rs`) -> `prover` -> `nova`
- `prover` -> `nova`
- `tools/src/command/jolt.rs` -> `jolt`
- `api` -> `prover` -> `nova` 

where the last has to take care to avoid a lot of coupling between the CLI and the `prover` crate. These pipelines are often independent from each other when they should be coherent, or intertwined with each other when they should be separated. This is not so big a deal at present, but in the short-term make it difficult to cleanly implement a future-proofed (as much as possible) SDK, and in the long-term will get unwieldy over time as various frontends and backends proliferate.

This PR is an initial move to a cleaner approach, where the `api` crate concentrates various backends that can be consumed by the various frontends (for now the CLI, Rust SDK, and `network` crate, in the future more network functionality as well as SDKs for other languages and environments).

In practice, this PR does three major things:

1. It moves the `config` and `prover` crates under the `api` crate, developing a single interface for frontends to access.
2. It slims down the `prover` crate, moving all CLI-dependent functionality into the `tools` crate and removing the vestigal SDK that was implemented within it.
3. It passes the Jolt implementation through the `api/prover` crate as well, again with the goal of having a single interface for frontends to access.

Also relevant is https://github.com/nexus-xyz/nexus-zkvm/issues/169. At present, the API crate is still "opinionated" about configuration of generic parameters, e.g., choices of curve pairings for Nova, random oracles, and commitment schemes. In the future we may want to let the various frontends to configure these generic parameters, which would require making the code in `api` suitably generic for the frontends to consume as needed.

### CODE REVIEW

File: config/Cargo.toml
Comment (slumber) [inline config/Cargo.toml:13]: config crate is used for various workspace configurations, and should be minimal.

An example is I want to add configs for RPC, like disconnect timeout, connections limit etc. With you PR I need to depend on `api` which bring, for example, spartan.

Work towards sdk, as I imagine it, should 
1. introduce `nexus-sdk` crate with generic trait interfaces
2. migrate workspace to using it

File: api/examples/prover_run.rs
Comment (mx00s) [inline api/examples/prover_run.rs:50]: Is there perhaps an off-by-one error here?

Since `..` is exclusive at the end, this executes one time less than the number of blocks in the trace. If we want one-call per block you could change the iterator's start value from `1` to `0`. Alternatively, there's `..=` which is inclusive at the end.

File: api/src/prover/nova/circuit.rs
Comment (mx00s) [inline api/src/prover/nova/circuit.rs:25]: 👍

File: tools/src/command/prove.rs
Comment (mx00s) [inline tools/src/command/prove.rs:207]: Can `vs.len()` ever be odd here? If so, the `ab[1]` indexing will sometimes panic because [slice's `chunks` method](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks) the last chunk returned may be shorter than the specified length when there's a remainder.

File: api/examples/prover_run.rs
Comment (sjudson) [reply api/examples/prover_run.rs:50]: The initial invocation at L45 handles the first step.

This function call should not be user facing (it'll be hidden behind the SDK and the CLI), but thinking about it it's probably worth introducing a more misuse resistant variant for when we don't need to hook in telemetry (like here), so I'll add that.

File: tools/src/command/prove.rs
Comment (sjudson) [reply tools/src/command/prove.rs:207]: Ah good call. Used correctly it always is, this is another artifact of losing misuse-resistance in order to move to a step-by-step interface to unhook the telemetry. I'll add an additional check.

File: config/Cargo.toml
Comment (sjudson) [reply config/Cargo.toml:13]: I'm amenable to moving the `config` crate back out, but at that point would say I think it makes more sense to move the configuration logic to the various crates its components refer to, rather than to have this small crate dangling off to the side from the `api` also trying to combine the various functionalities spread around into one coherent interface. But if you feel strongly about keep `config` where and how it is I can move it back. 

As for the proposed flow, I agree in principle. The problem in practice is that the `prover` crate is so tied into the CLI at present that I can't make a reasonable, generic SDK that uses it in the way I'd like without doing this restructuring first.

**Comment by sjudson** [comment]
Restored the original `config` crate at @slumber's request.

File: api/Cargo.toml
Comment (slumber) [inline api/Cargo.toml:10]: This file needs a cleanup, e.g. I don't think it should depend on `tui` and have `verbose` feature anymore

**Comment by slumber** [comment]
As for the sdk shape: I imagined we could start with very minimal 
```rust
// call it nexus-prover-core crate or smth
trait Prover {}
```
which should be as lightweight as possible

Then, all prover crates would depend on it.
```rust
// nexus-nova
impl Prover for NovaProver {}
// nexus-jolt
impl Prover for JoltProver {}
```
Finally,
```rust
// nexus-sdk
pub use Prover;
#[cfg(feature = "nova")]
pub use nexus_nova::NovaProver;
#[cfg(feature = "jolt")]
pub use nexus_jolt::JoltProver;
```

Maybe not so much configurable in terms of curves/fields/k/etc., perhaps snips some prover-specific features, but clean and easy to use.

**Comment by sjudson** [comment]
Converted back to draft to enable some other relevant code to come through first, that this will be adapted to.

**Comment by mx00s** [comment]
> Thoughts on this @mx00s?

To start, I'm going to avoid using the terms API and SDK.

I agree with the idea of leveraging clearly defined interfaces to improve cross-team coordination.

> I've worked in a context before where you have a proliferation of frontend SDKs/CLI tooling and backends, and the only reason it was manageable from a cross-team coordination standpoint was because the web API that linked everything together provided a single, concrete interface that enabled clear interoperability.

That will be necessary as 

1. we iteratively improve implementations of those interfaces,
2. create new implementations, and
3. support a growing variety of consumers.

I also appreciate the benefits of designing simple interfaces starting top-down as opposed to bottom-up as @slumber seems to suggest:

> going back to review https://github.com/nexus-xyz/nexus-zkvm/pull/170#discussion_r1633268126, I don't understand the approach of rewriting anything before the actual sdk is implemented. Meaning: come up with generic trait interface (https://github.com/nexus-xyz/nexus-zkvm/issues/148 original description seems to be ignored), introduce it as a crate into the workspace, only then start migrating code like tools or vm to use these interfaces.

In terms of API and SDK dichotomy as it's been presented so far, my understanding is the API is meant to be a lower-level abstraction for IVC and the SDK will eventually depend on it, providing higher-level interfaces that encapsulate more of the details. I can imagine this working well, but admittedly am hazy on who the intended consumers of the SDK would be and what details relevant to the API they'd benefit from having encapsulated.

It's abundantly clear we need clean, simple interfaces to support the aforementioned coordination points over time. Whether to design such interfaces working bottom-up or top-down is somewhat a matter of taste, but I am admittedly biased toward the top-down approach to optimize for ergonomic UX. Admittedly, sometimes the top-down approach means discovering while wiring up the internals that the high-level interfaces must be adjusted, but that's okay.

I've mentioned it elsewhere, but if we align on the top-down approach I think we'd benefit from:

1. implementing end-to-end tests https://github.com/nexus-xyz/nexus-zkvm/issues/165
2. defining high-level interfaces that make those tests clean
3. promoting those high-level interfaces to a public crate that serves as a coordination point, whatever we choose to call it

By the way, like @sjudson I don't see "api" as usually pertaining to networking. I suppose OpenAPI/Swagger may have shifted norms in that direction a bit.

**Comment by sjudson** [comment]
Removing from draft due to merge of https://github.com/nexus-xyz/nexus-zkvm/pull/185

File: config/Cargo.toml
Comment (slumber) [inline config/Cargo.toml:24]: it's best to checkout `/config/*` to match main

Comment (slumber) [reply config/Cargo.toml]: My comment is about the whole directory, notice there're toml files recommitted as well as a backup mod.rs

Comment (sjudson) [reply config/Cargo.toml]: Ugh, I did checkout the whole directory, not sure why git didn't clear deleted files...

File: network/rpc/server/Cargo.toml
Comment (slumber) [reply network/rpc/server/Cargo.toml:24]: Notice how bringing `nexus-api` here also requires compiling jolt as a dependency, while it's unused.

Logically one would feature-gate api modules, but then either all of these features get propagated into depending crate, or disabled by default, making development experience worse.

File: tools/src/lib.rs
Comment (slumber) [inline tools/src/lib.rs:50]: "verbose" feature is not defined as the cli is always verbose. `tools` use `TerminalHandle::new_enabled()`

File: jolt/src/lib.rs
Comment (sjudson) [reply jolt/src/lib.rs:62]: As discussed, goal here is to have a single interface that re-exports necessary functionality for various frontends/network components to rely upon.

Comment (slumber) [reply jolt/src/lib.rs:62]: I get that api re-exports, but this isn't re-exporting, this is moving jolts logic into a different crate, scattering it over the workspace.

Meaning, if I look into `nexus-jolt` code, there's code for translating trace, but no code for generating actual proof. So even if I want to use it bypassing api, I can't.

All jolt methods are public, why is the following not sufficient?
```rust
// nexus-api
pub use nexus_jolt::*;
```

Comment (sjudson) [reply jolt/src/lib.rs:62]: I see, yes, I can move in that direction.

File: network/rpc/server/Cargo.toml
Comment (sjudson) [reply network/rpc/server/Cargo.toml]: Fixed.

File: tools/src/lib.rs
Comment (sjudson) [reply tools/src/lib.rs]: Fixed.

File: jolt/src/lib.rs
Comment (sjudson) [reply jolt/src/lib.rs]: Fixed.

**Comment by slumber** [comment]
I assume https://github.com/nexus-xyz/nexus-zkvm/pull/170#discussion_r1645049397 is not going to be addressed. 

Given the amount of design-discussions about _potential issues_ we will face in years after implementing SDK, and that this PR already makes it worse to build on top of workspace crates, I'll again outline how it just doesn't work.

My impression is it's crafted to facilitate `tools` which never needed it.

Please get an approval from @mx00s to unblock the merge.

**Comment by sjudson** [comment]
After some further discussion, added feature gating @mx00s @slumber: https://github.com/nexus-xyz/nexus-zkvm/pull/170/commits/fde39c00a8df8ac85d0a1df36eafeda9da834b60

**Comment by sjudson** [comment]
Updated dependency graph:

image

replacing 

image

**Comment by mx00s** [comment]
Nice. For reference, the following command can be used to produce those dependency graphs:

```bash
cargo install cargo-depgraph
cargo depgraph --workspace-only | dot -Tpng > graph.png
```

**Review by mx00s — APPROVED** [review]
With the dependency graph reconfigured, we're better positioned to iterate on `nexus-api` and support the many-to-many dependence between prover frontends and backends.

Thanks for adding the prover feature flags too. This should help as continue implementing/improving provers and even vary defaults per deployment environment, for instance.

LGTM!

**Pull request merged**
